### PR TITLE
prov/cxi: Initial upstreaming of the CXI provider

### PR DESCRIPTION
**Pull request by jswaro**
This is the squashed commit of the CXI provider for upstreaming to OFIWG. This is the combined work of multiple authors, all of which have been given attribution on the initial commit. 

This commit contains the CXI provider, man pages, and criterion tests needed for baseline validation. Automated testing is not yet set up and will be handled in the coming weeks.

### CODE REVIEW

File: prov/cxi/configure.m4 [hunk 1]
```diff
@@ -0,0 +1,153 @@
+dnl SPDX-License-Identifier: GPL-2.0
+dnl
+dnl Copyright 2018 Cray Inc. All rights reserved.
+
```
Comment (muttormark) [inline prov/cxi/configure.m4:4]: Do we need to augment the copyright here?

File: prov/cxi/include/cxip_faults.h [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+}
+static inline bool cxip_trap_search(int index, int trap, int *err)
+{
+	return _cxip_trap_search(index, trap, err);
+}
+#else
+static inline void cxip_trap_close(void) {}
+static inline void cxip_trap_set(int a, int b, int c) {}
+static inline bool cxip_trap_search(int a, int b, int *c) {return false;}
+#endif
```
Comment (muttormark) [inline prov/cxi/include/cxip_faults.h:148]: newline?

File: man/man7/fi_cxi.7 [hunk 3]
```diff
@@ -0,0 +1,2144 @@
+.\"t
```
Comment (swelch) [inline man/man7/fi_cxi.7:1]: I believe this file should get generated automatically when upstream.

Comment (jswaro) [reply man/man7/fi_cxi.7:1]: The build failed when this wasn't present. I included it, but I can remove it.

Comment (swelch) [reply man/man7/fi_cxi.7:1]: Maybe there is a chicken and egg issue here. Possibly the bot to create the man page output needs to be kicked off first this initial time.

**Comment by zachdworkin** [comment]
Please rebase to pick up changes from #9762 to get intel-ci to pass

File: prov/cxi/include/cxip_faults.h [hunk 4]
```diff
@@ -0,0 +1,148 @@
+/*
+ * SPDX-License-Identifier: GPL-2.0
+ *
+ * Copyright (c) 2019 Hewlett Packard Enterprise Development LP
+ */
+
```
Comment (j-xiong) [inline prov/cxi/include/cxip_faults.h:6]: Missing  `#ifndef ... #define ...  #endif` protection?

**Comment by j-xiong** [comment]
Are `cassini-headers` , `cxi-uapi-headers`, and `libcxi` available as packages for ubuntu-20.4 and 22.4? Just want to check if it is possible to add cxi to the build check workflow.

File: prov/cxi/configure.m4 [hunk 5]
```diff
@@ -46,11 +46,11 @@ AC_DEFUN([FI_CXI_CONFIGURE],[
 		[
 			AC_CHECK_HEADER(cxi_prov_hw.h,
 				[],
-				[AC_MSG_ERROR("cxi_prov_hw.h not found")])
+				[cxi_happy=0])
```
Comment (j-xiong) [inline prov/cxi/configure.m4:49]: Please squash this into the first commit.

Comment (j-xiong) [reply prov/cxi/configure.m4:49]: I mean the entire commit, not this specific change.

File: man/fi_cxi.7.md [hunk 6]
```diff
@@ -0,0 +1,1782 @@
...
+    with ATS to help avoid network page faults. Valid values are "off" and
+    "all". When mlock mode is "off", the provider does not use mlock(). An
+    application using ATS without mlock() may experience network page faults,
+    reducing network performance. When ats_mlock_mode is set to "all", the
+    provider uses mlockall() during initialization with ATS. mlockall() causes
+    all mapped addresses to be locked in RAM at all times. This helps to avoid
+    most network page faults. Using mlockall() may increase pressure on
+    physical memory.  Ignored when ODP is disabled.
+
+*FI_CXI_RDZV_THRESHOLD*
```
[the remark this answers is not in the document]
Comment (swelch) [reply man/fi_cxi.7.md:981]: I understand how it would be nice to use consistent naming, but this provider is already being used and many environments are being setup using this naming. We could always map both to the same value, but I personally think that creates more confusion from a users perspective.

**Comment by jswaro** [comment]
> Are `cassini-headers` , `cxi-uapi-headers`, and `libcxi` available as packages for ubuntu-20.4 and 22.4? Just want to check if it is possible to add cxi to the build check workflow.

Not yet.

File: man/fi_cxi.7.md [hunk 7]
```diff
@@ -0,0 +1,1782 @@
...
+The provider uses the libCXI library for control operations and a set of
+Cassini-specific header files to enable direct hardware access in the data path.
+
+# SUPPORTED FEATURES
+
+The CXI provider supports a subset of OFI features.
+
+## Endpoint types
+
+The provider supports the *FI_EP_RDM* endpoint type, including scalable
```
[the remark this answers is not in the document]
Comment (jswaro) [reply man/fi_cxi.7.md:39]: @iziemba , @swelch : Is support present or not? L300 conflicts with L39.

File: man/fi_cxi.7.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (jswaro) [reply man/fi_cxi.7.md:981]: I would like to consider this as a future refinement. I understand where you are coming from @j-xiong , but we will need time to transition users to a new environment variable and deprecate this one over a period of time.

File: prov/cxi/include/cxip_faults.h (hunk 4 above)
Comment (jswaro) [reply prov/cxi/include/cxip_faults.h:6]: There is an internal PR that I'll pull out which address that. It was discovered with RHEL 9.3 testing. 

A note for us though -- RHEL 9.3 default LDFLAGS includes -flto=auto, which breaks/strips symbols from the static library generated during RPM builds. This doesn't affect local builds (configure -> make -> make install) because the RPM macros from RHEL aren't picked up in that workflow.

File: prov/cxi/configure.m4 (hunk 5 above)
Comment (jswaro) [reply prov/cxi/configure.m4:49]: I plan to. I want to gather as much feedback as possible without resquashing and losing comments (it happens sometimes).

File: man/fi_cxi.7.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (swelch) [reply man/fi_cxi.7.md:39]: @jswaro The current version does not, this reference to SEP should be removed.

File: Makefile.am [hunk 8]
```diff
@@ -445,7 +445,7 @@ test:
 	./util/fi_info
 
 rpm: dist
-	LDFLAGS=-Wl,--build-id rpmbuild -ta libfabric-$(PACKAGE_VERSION).tar.bz2
+	LDFLAGS=-Wl,--build-id rpmbuild -ta $(RPMBUILD_OPTS) libfabric-$(PACKAGE_VERSION).tar.bz2
```
Comment (shefty) [inline Makefile.am:448]: nit: This specific change is unrelated to the provider and should be a separate patch.

File: prov/cxi/Makefile.include [hunk 9]
```diff
@@ -0,0 +1,211 @@
+# SPDX-License-Identifier: GPL-2.0
```
Comment (shefty) [inline prov/cxi/Makefile.include:1]: This license is not aligned with the current codebase.

**Review by shefty — CHANGES_REQUESTED** [review]
License in several files is not aligned with the project.

File: Makefile.am (hunk 8 above)
Comment (jswaro) [reply Makefile.am]: I'll pull it out

**Comment by jswaro** [comment]
@shefty : Licenses have been fixed.

**Comment by shefty** [comment]
@jswaro - please merge the commit that updates the license with the original commit, so git bisect can't pick up an incorrect license.  Thanks for updating!

**Comment by jswaro** [comment]
Everything has been squashed.

**Comment by jswaro** [comment]
@j-xiong , @shefty , @iziemba : Any last requests for this PR?

File: prov/cxi/include/cxip_faults.h (hunk 4 above)
Comment (j-xiong) [reply prov/cxi/include/cxip_faults.h:6]: Are you going to add the `#ifdef` here? It's a simple change.

**Review by swelch — APPROVED** [review]

**Review by shefty — APPROVED** [review]
